plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -130,6 +229,127 @@ class bp_connection_manager {
+               } else if (prev->producer_name > peer.producer_name) {
+                  return false; // required to be sorted
+               }
+            }
+            if (++producers[peer.producer_name] > 4)
+               return false; // only 4 entries per producer allowed
+            prev = &peer;
+         }
+      }
+
+      controller& cc = self()->chain_plug->chain();
+
+      fc::lock_guard g(gossip_bps.mtx);
+      auto& sig_idx = gossip_bps.index.get<by_sig>();
+      for (auto i = msg.peers.begin(); i != msg.peers.end();) {
+         const auto& peer = *i;
+         try {
+            if (!sig_idx.contains(peer.sig)) { // we already have it, already verified
+               // peer key may have changed or been removed, so if invalid or not found skip it
+               if (std::optional<public_key_type> peer_key = cc.get_peer_key(peer.producer_name)) {
+                  public_key_type pk(peer.sig, peer.digest());
+                  if (pk != *peer_key) {
+                     fc_dlog(self()->get_logger(), "Recovered peer key did not match on-chain ${p}, recovered: ${pk} != expected: ${k}",
+                             ("p", peer.producer_name)("pk", pk)("k", *peer_key));
+                     i = msg.peers.erase(i);
+                     continue;
+                  }
+               } else { // unknown key
+                  fc_dlog(self()->get_logger(), "Failed to find peer key ${p}", ("p", peer.producer_name));
+                  i = msg.peers.erase(i);
+                  continue;
+               }
+            }
+         } catch (fc::exception& e) {
+            fc_dlog(self()->get_logger(), "Exception recovering peer key ${p}, error: ${e}", ("p", peer.producer_name)("e", e.to_detail_string()));
+            // invalid key
+            i = msg.peers.erase(i);
+            continue;
+         }
+         ++i;
+      }
+
+      return !msg.peers.empty();
+   }
+
+   // thread-safe
+   bool update_gossip_bps(const gossip_bp_peers_message& msg) {
+      // providing us with full set
+      fc::unique_lock g(gossip_bps.mtx);
+      auto& idx = gossip_bps.index.get<by_producer>();
+      bool diff = false;
+      for (const auto& peer : msg.peers) {
+         if (auto i = idx.find(boost::make_tuple(peer.producer_name, boost::cref(peer.server_address))); i != idx.end()) {
+            if (*i != peer) {
+               gossip_bps.index.modify(i, [&peer](auto& m) {
+                  m = peer;
+               });
+               diff = true;
+            }
+         } else {
+            gossip_bps.index.insert(peer);
+            diff = true;
+         }
+      }
+      g.unlock();
+      if (diff) {
+         connect_to_active();
+      }
+      return diff;
+   }
+
+   // thread-safe
+   void connect_to_active() {
+      // do not hold mutexes when calling resolve_and_connect which acquires connections mutex since other threads
+      // can be holding connections mutex when trying to acquire these mutexes
```

Comment:
We do hold a mutex in `on_pending_schedule` though? Is it not an issue there?
Also, can we avoid this code duplication?

Reply:
Yes, that is an issue. Thanks.